src/lib/message/L2ToL1Message.ts

```diff
@@ -117,94 +49,104 @@ export enum L2ToL1MessageStatus {
+  protected isClassic(
+    e: L2ToL1TransactionEvent
+  ): e is EventArgs<ClassicL2ToL1TransactionEvent> {
+    return isDefined(
+      (e as EventArgs<ClassicL2ToL1TransactionEvent>).indexInBatch
+    )
+  }
 
   public static fromEvent<T extends SignerOrProvider>(
     l1SignerOrProvider: T,
-    event: EventArgs<L2ToL1TxEvent>
+    event: L2ToL1TransactionEvent
   ): L2ToL1MessageReaderOrWriter<T>
-  public static fromEvent<T extends SignerOrProvider>(
+  static fromEvent<T extends SignerOrProvider>(
     l1SignerOrProvider: T,
-    event: EventArgs<L2ToL1TxEvent>
+    event: L2ToL1TransactionEvent
   ): L2ToL1MessageReader | L2ToL1MessageWriter {
     return SignerProviderUtils.isSigner(l1SignerOrProvider)
       ? new L2ToL1MessageWriter(l1SignerOrProvider, event)
       : new L2ToL1MessageReader(l1SignerOrProvider, event)
   }
 
+  /**
+   * Get event logs for L2ToL1 transactions.
+   * @param l2Provider
+   * @param filter Block range filter
+   * @param position The batchnumber indexed field was removed in nitro and a position indexed field was added.
+   * For pre-nitro events the value passed in here will be used to find events with the same batchnumber.
+   * For post nitro events it will be used to find events with the same position.
+   * @param destination The L1 destination of the L2ToL1 message
+   * @param hash The uniqueId indexed field was removed in nitro and a hash indexed field was added.
+   * For pre-nitro events the value passed in here will be used to find events with the same uniqueId.
+   * For post nitro events it will be used to find events with the same hash.
+   * @param indexInBatch The index in the batch, only valid for pre-nitro events. This parameter is ignored post-nitro
+   * @returns Any classic and nitro events that match the provided filters.
+   */
   public static async getL2ToL1Events(
     l2Provider: Provider,
     filter: { fromBlock: BlockTag; toBlock: BlockTag },
     position?: BigNumber,
     destination?: string,
-    hash?: BigNumber
-  ): Promise<(EventArgs<L2ToL1TxEvent> & { transactionHash: string })[]> {
-    const eventFetcher = new EventFetcher(l2Provider)
+    hash?: BigNumber,
+    indexInBatch?: BigNumber
+  ): Promise<(L2ToL1TransactionEvent & { transactionHash: string })[]> {
     return (
-      await eventFetcher.getEvents(
-        ArbSys__factory,
-        t => t.filters.L2ToL1Tx(null, destination, hash, position),
-        { ...filter, address: ARB_SYS_ADDRESS }
-      )
-    ).map(l => ({ ...l.event, transactionHash: l.transactionHash }))
+      await Promise.all([
+        classic.L2ToL1MessageClassic.getL2ToL1MessageLogs(
+          l2Provider,
+          filter,
+          position,
+          destination,
+          hash,
+          indexInBatch
+        ),
+        nitro.L2ToL1MessageNitro.getL2ToL1Events(
+          l2Provider,
+          filter,
+          position,
+          destination,
+          hash
+        ),
```

Comment:
in a future PR we could prob make this more efficient by reducing the number of log queries

Reply:
yeh, my intuition for a standard index is that this would actually make the query slower as it would introduce OR terms, that standard indexs struggle with. Although ethereum indexes are nothing like standard ones, so I dont know if that applies here. 

One simple thing we can defo do is look at the from/to block and compare to the nitro activation block to adjust the ranges, whilst this will still be two network calls it should be less work for the node. Added that here.